pkg/gadgets/top/ebpf/piditer/bpf/pid_iter.bpf.c

```diff
@@ -0,0 +1,38 @@
+// SPDX-License-Identifier: (GPL-2.0-only OR BSD-2-Clause)
+/* Copyright (c) 2020 Facebook, (c) 2022 The Inspektor Gadget authors */
+#include <vmlinux/vmlinux.h>
+#include <bpf/bpf_helpers.h>
+#include <bpf/bpf_core_read.h>
+#include <bpf/bpf_tracing.h>
+#include "pid_iter.h"
+
+const volatile __u64 bpf_prog_fops_addr = 0;
+
+struct pid_iter_entry *unused __attribute__((unused));
+
+SEC("iter/task_file")
+int gadget_ebpftop(struct bpf_iter__task_file *ctx)
+{
+	struct file *file = ctx->file;
+	struct task_struct *task = ctx->task;
+	struct pid_iter_entry e;
+
+	if (!file || !task)
+		return 0;
+
+	// If we do have an address of bpf_prog_fops, use it; otherwise... well...
```

Comment:
They say "there be dragons".

Seriously though:
If `bpf_prog_fops_addr` is 0, the program will run on more than just programs. In tests it worked and didn't have any weird outcome (`BPF_CORE_READ` is the only action specific to bpf programs and it will return 0 if it accesses invalid memory. But I can imagine that it also potentially could return garbage if the address of `file->private_data` is valid and contains enough data - that would lead to a wrong program id and possibly result in ebpf programs showing wrong pid/comm information).

This was implemented in the first version where we had no fallback for the iterator. Now that we have one, it might be better to entirely skip the ebpf iterator and fallback to the /proc/./fdinfo scanning when the address of `bpf_prog_fops_addr` can't be found.

Reply:
> Now that we have one, it might be better to entirely skip the ebpf iterator and fallback to the /proc/./fdinfo scanning when the address of bpf_prog_fops_addr can't be found.

 I agree. `getBpfProgFopsAddr()` can return an error, then it can be checked on `NewTracer()`.